libevmasm/RuleList.h

```diff
@@ -597,6 +626,24 @@ std::vector<SimplificationRule<Pattern>> simplificationRuleListPart7(
 		}
 	});
 
+	rules.push_back({
+		Builtins::SHL(A, Builtins::SIGNEXTEND(B, X)),
+		[=]() -> Pattern { return Builtins::SIGNEXTEND(A.d() / 8 + B.d(), Builtins::SHL(A, X)); },
+		[=] { return A.d() % 8 == 0 && A.d() <= 256 && B.d() <= 32; }
+	});
```

Comment:
Yes, it pulls signextend across a shift. We do the same with AND. It can be combined with https://github.com/ethereum/solidity/pull/11748/files#diff-197806d78c74566018771b7942feb91b40c5ee67ad20ad13f955dad525c6d5f8R337 to remove the signextend even if `A / 8 + B` is small.

This step is useful for the conversion `int8` -> `bytes1`, where we first `signextend`, then `shift` and finally `mask`.
These are all reduced to a single `shift`. I can add a specific test for that.